src/client/pydaos/torch/torch_shim.c

```diff
@@ -0,0 +1,887 @@
+		return NULL;
+	}
+
+	Py_buffer bview;
+	if (PyObject_GetBuffer(buffer, &bview, PyBUF_WRITE) == -1) {
+		PyErr_SetString(PyExc_BufferError, "Buffer is not writable");
+		return NULL;
+	}
+
+	/*
+	  Since python can use buffer like objects that might not have contiguous memory layout,
+	  let's put a guardrail accepting only buffers with contiguous memory region
+	*/
+	if (!PyBuffer_IsContiguous(&bview, 'C')) {
+		PyErr_SetString(PyExc_BufferError, "Buffer is not contiguous");
+		PyBuffer_Release(&bview);
+		return NULL;
+	}
+
+	daos_size_t read = bview.len;
+
+	d_iov_t     iov;
+	d_iov_set(&iov, bview.buf, read);
+
+	d_sg_list_t sgl = {
+	    .sg_nr     = 1,
+	    .sg_nr_out = 0,
+	    .sg_iovs   = &iov,
+	};
+
+	rc = dfs_lookup(hdl->dfs, path, O_RDONLY, &obj, NULL, NULL);
+	if (rc) {
+		D_ERROR("Could not lookup '%s': %s (rc=%ld)", path, strerror(rc), rc);
+		rc = -rc;
+		goto out;
+	}
+
+	rc = dfs_read(hdl->dfs, obj, &sgl, 0 /* offset */, &read, NULL);
+	if (rc) {
+		rc = -rc;
+		goto out;
+	}
+	rc = read;
+
+out:
+	PyBuffer_Release(&bview);
+
+	if (obj) {
+		int rc2 = dfs_release(obj);
+		if (rc2) {
+			D_ERROR("Could not release object '%s': %s (rc=%d)", path, strerror(rc2),
+				rc2);
+		}
+	}
+
+	return PyLong_FromLong(rc);
+}
+
+/* describes the IO operation */
+struct io_op {
+	daos_event_t ev;
+
+	dfs_obj_t   *obj;
+	PyObject    *item;
+
+	/* Purely for debug purpose: should not be freed as it's not the owner of the data */
+	const char  *path;
+
+	daos_size_t  size;
+	int          err;
+
+	d_iov_t      iov;
+	d_sg_list_t  sgl;
+
+	/* Buffer view that implements Python Buffer Protocol, has to be release after use */
+	Py_buffer    buf_view;
+};
+
+static int
+start_read_op(struct dfs_handle *hdl, PyObject *item, struct io_op *op)
+{
+	assert(op != NULL);
+
+	int           rc  = 0;
+	daos_event_t *evp = &op->ev;
+
+	PyObject     *py_path = PyTuple_GetItem(item, 0);
+	PyObject     *py_buff = PyTuple_GetItem(item, 1);
+
+	if (py_path == NULL || py_buff == NULL) {
+		D_WARN("Each tuple must contain exactly two elements: path and bytearray");
+		return EINVAL;
+	}
+
+	const char *path = PyUnicode_AsUTF8(py_path);
+	if (path == NULL) {
+		D_WARN("First element of a tuple does not look like a path");
+		return EINVAL;
+	}
+
+	if (PyObject_GetBuffer(py_buff, &op->buf_view, PyBUF_WRITE) == -1) {
+		D_WARN("Buffer is not writable");
+		return EINVAL;
+	}
+
+	if (!PyBuffer_IsContiguous(&op->buf_view, 'C')) {
+		D_WARN("Buffer for '%s' is not contiguous", path);
+		rc = EINVAL;
+		goto out;
+	}
+
+	rc = daos_event_init(evp, hdl->eq, NULL);
+	if (rc) {
+		D_ERROR("Could not init event: %s (rc=%d)", d_errstr(rc), rc);
+		goto out;
+	}
+
+	rc = dfs_lookup(hdl->dfs, path, O_RDONLY, &op->obj, NULL, NULL);
+	if (rc) {
+		D_ERROR("Could not lookup path '%s': %s (rc=%d)", op->path, strerror(rc), rc);
+		goto out;
+	}
+
+	op->path = path;
+	op->size = op->buf_view.len;
+	d_iov_set(&op->iov, op->buf_view.buf, op->size);
```

Comment:
i assume we don't know the file size at this point, right?

Reply:
@johannlombardi we do, the `op->buf_view.len` has the expected size of the file.  The buffer itself was allocated on the python side before the calling the shim layer.